File: Programmers/Lv_2/PCCP_기출문제_2번_석유_시추/PCCP_기출문제_2번_석유_시추_지민성.java
import java.util.*;

class Solution {
    
    public static boolean[][] visited;
    public static int row;
    public static int col;
    
    public static int[] dx = {1, 0, -1, 0};
    public static int[] dy = {0, 1, 0, -1};
    public static int index = 1;
    public static Map<Integer, Integer> dic = new HashMap<>();
    
    public int solution(int[][] land) {
        
        int answer = 0;
        row = land.length;
        col = land[0].length;
        visited = new boolean[row][col];
        
        for (int i = 0; i < row; i++)
            for (int j = 0; j < col; j++)
                if (land[i][j] == 1)
                    bfs(land, i, j);
        for (int j = 0; j < col; j++) {
            Set<Integer> result = new HashSet<>();
            int temp = 0;
            for (int i = 0; i < row; i++)
                if (land[i][j] != 0)
                    result.add(land[i][j]);
            for (int i : result)
                temp = temp + dic.get(i);                
            answer = Math.max(answer, temp);
        }
        return answer;
    }
    
    public void bfs(int[][]land, int i, int j) {
        Queue<Node> queue = new LinkedList<>();
        Node node = new Node(i, j);
        queue.offer(node);
        index++;
        int count = 1;
        visited[i][j] = true;
        land[i][j] = index;
        
        while(!queue.isEmpty()) {
            Node current = queue.poll();
            int x = current.r;
            int y = current.c;
            
            for (int d = 0; d < 4; d++) {
                int nx = x + dx[d];
                int ny = y + dy[d];
                if (nx < 0 || nx >= row || ny < 0 || ny >= col)
                    continue;
                if (visited[nx][ny] || land[nx][ny] == 0)
                    continue;
                Node newNode = new Node(nx, ny);
                queue.offer(newNode);
                visited[nx][ny] = true;
                land[nx][ny] = index;
                count++;
            }
        }
        dic.put(index, count); 
    }
    
    public class Node {
        int r;
        int c;
        public Node(int r, int c) {
            this.r = r;
            this.c = c;
        }
    }
}

File: Programmers/Lv_2/PCCP_기출문제_2번_석유_시추/PCCP_기출문제_2번_석유_시추_지민성.py
from collections import deque, defaultdict


def solution(land):
    row, col = len(land), len(land[0])
    dx, dy = [1, 0, -1, 0], [0, 1, 0, -1]
    visited = [[False] * col for _ in range(row)]
    dic = defaultdict(int)
    order = 0
    for i in range(row):
        for j in range(col):
            if land[i][j] == 1 and not visited[i][j]:
                order = order + 1
                stack = []
                count = 1
                queue = deque()
                queue.append([i, j])
                stack.append([i, j])
                visited[i][j] = True
                while queue:
                    x, y = queue.popleft()
                    for d in range(4):
                        nx = x + dx[d]
                        ny = y + dy[d]
                        if nx < 0 or nx >= row or ny < 0 or ny >= col:
                            continue
                        if visited[nx][ny] or land[nx][ny] == 0:
                            continue
                        queue.append([nx, ny])
                        stack.append([nx, ny])
                        count = count + 1
                        visited[nx][ny] = True
                for x, y in stack:
                    land[x][y] = order
                dic[order] = count
    result = 0
    for i in range(col):
        temp_set = set()
        for j in range(row):
            temp_set.add(land[j][i])
        temp = 0
        for idx in temp_set:
            temp = temp + dic[idx]
        result = max(result, temp)
    return result

print(solution([[0, 0, 0, 1, 1, 1, 0, 0], [0, 0, 0, 0, 1, 1, 0, 0], [1, 1, 0, 0, 0, 1, 1, 0], [1, 1, 1, 0, 0, 0, 0, 0], [1, 1, 1, 0, 0, 0, 1, 1]]))
print(solution([[1, 0, 1, 0, 1, 1], [1, 0, 1, 0, 0, 0], [1, 0, 1, 0, 0, 1], [1, 0, 0, 1, 0, 0], [1, 0, 0, 1, 0, 1], [1, 0, 0, 0, 0, 0], [1, 1, 1, 1, 1, 1]]))
